### PR TITLE
Rename both scope and var_type of anonymous types

### LINKED ISSUE #3717: non-deterministic anon_struct
While working on [reproducible builds](https://reproducible-builds.org/) for [openSUSE](https://en.opensuse.org/openSUSE:Reproducible_Builds), I found that our `glfw` package varies in every build since the upgrade to geany-2.0

I extracted this minimal reproducer:

```bash
cat > test2.c <<EOF
typedef struct {
  struct
  {
  } __value32;
} __atomic_wide_counter;
EOF
for i in 1 2 ; do geany -g glfw.c.tags test2.c ; md5sum glfw.c.tags ; done | sort -u | wc -l
2
```

The output differs thusly:
```diff
-__value32�64�__atomic_wide_counter�0�__anonceed23430108::anon_struct_1
+__value32�64�__atomic_wide_counter�0�__anone35807a60108::anon_struct_1
```

I think, there is the same problem with anon unions.

Please make output deterministic.

### PR DESCRIPTION
**Pull request by techee**
In addition to scope, anonymous type can appear also inside var_type which should be renamed as well.

Fixes #3717.

### CODE REVIEW

**Comment by bmwiedemann** [comment]
I tested our `glfw` build with this patch and it became perfectly reproducible + deterministic.
Thanks a lot.

**Comment by bmwiedemann** [comment]
When will this be merged?

**Comment by techee** [comment]
@b4n Does it look OK to you?

We have to rename the anonymous tag name both in scope and var_type so I moved the renaming to a separate function and call the function both on scope and var_type.

File: src/tagmanager/tm_ctags.c [hunk 1]
```diff
@@ -241,6 +241,29 @@ void tm_ctags_clear_ignore_symbols(void)
 }
 
 
+static void replace_str(gchar **where, const gchar *what, guint what_len,
+	const gchar *replacement, guint replacement_len)
```
Comment (b4n) [inline src/tagmanager/tm_ctags.c:245]: Just a remark, don't change anything for me here: this "API" is quite terrible (not all string args have a corresponding length, and they are still not allowed to lack a terminator anyway), but I guess it makes sense for how it's used (saves some `strlen()`s if the anon name is used more than once).

File: src/tagmanager/tm_ctags.c [hunk 2]
```diff
@@ -361,22 +383,14 @@ static void rename_anon_tags(TMSourceFile *source_file)
...
-					strncpy(str, nested_tag->scope, prefix_len);
-					strcpy(str + prefix_len, new_name);
-					strcpy(str + prefix_len + new_name_len, pos + orig_name_len);
-					g_free(nested_tag->scope);
-					nested_tag->scope = str;
-				}
+				replace_str(&nested_tag->scope, orig_name, orig_name_len, new_name, new_name_len);
+
+				/* Do the same for var_type as well */
+				replace_str(&nested_tag->var_type, orig_name, orig_name_len, new_name, new_name_len);
```
Comment (b4n) [inline src/tagmanager/tm_ctags.c:393]: Shouldn't this new helper be used around L406?

File: src/tagmanager/tm_ctags.c [hunk 3]
```diff
@@ -241,6 +241,29 @@ void tm_ctags_clear_ignore_symbols(void)
...
+{
+	if (where && *where)
+	{
+		guint where_len = strlen(*where);
+		gchar *pos = strstr(*where, what);
+
+		if (pos)
+		{
+			gchar *str = g_malloc(where_len + replacement_len);
+			guint prefix_len = pos - *where;
```
Comment (b4n) [inline src/tagmanager/tm_ctags.c:255]: conversion: pointer difference yields signed (although we know it's `>= 0` in this case indeed), so we "need" an explicit cast

**Review by b4n — CHANGES_REQUESTED** [review]
Looks good, but impl can possibly be improved a tad (see nitpicks & useful comments inline).

Suggested changes (don't trust it blindly, esp. the part around L406):
```diff
diff --git a/src/tagmanager/tm_ctags.c b/src/tagmanager/tm_ctags.c
index ec795a31d..c7ed72fe0 100644
--- a/src/tagmanager/tm_ctags.c
+++ b/src/tagmanager/tm_ctags.c
@@ -241,26 +241,30 @@ void tm_ctags_clear_ignore_symbols(void)
 }
 
 
-static void replace_str(gchar **where, const gchar *what, guint what_len,
+static gboolean replace_str(gchar **where, const gchar *what, guint what_len,
 	const gchar *replacement, guint replacement_len)
 {
 	if (where && *where)
 	{
-		guint where_len = strlen(*where);
 		gchar *pos = strstr(*where, what);
 
 		if (pos)
 		{
-			gchar *str = g_malloc(where_len + replacement_len);
-			guint prefix_len = pos - *where;
+			gsize where_len = strlen(*where);
+			gchar *str = g_malloc(where_len + (replacement_len - what_len) + 1);
+			gsize prefix_len = (gsize) (pos - *where);
 
 			strncpy(str, *where, prefix_len);
 			strcpy(str + prefix_len, replacement);
 			strcpy(str + prefix_len + replacement_len, pos + what_len);
 			g_free(*where);
 			*where = str;
+
+			return TRUE;
 		}
 	}
+
+	return FALSE;
 }
 
 
@@ -399,22 +403,13 @@ static void rename_anon_tags(TMSourceFile *source_file)
 			{
 				TMTag *var_tag = TM_TAG(source_file->tags_array->pdata[j]);
 				guint var_scope_len = var_tag->scope ? strlen(var_tag->scope) : 0;
-				gchar *pos;
 
 				/* Should be at the same scope level as the anon tag */
-				if (var_scope_len == scope_len &&
-					var_tag->var_type && (pos = strstr(var_tag->var_type, orig_name)))
+				if (var_scope_len != scope_len || ! var_tag->var_type ||
+					! replace_str(&var_tag->var_type, orig_name, orig_name_len, new_name, new_name_len))
 				{
-					GString *str = g_string_new(var_tag->var_type);
-					gssize p = pos - var_tag->var_type;
-					g_string_erase(str, p, strlen(orig_name));
-					g_string_insert(str, p, new_name);
-					g_free(var_tag->var_type);
-					var_tag->var_type = str->str;
-					g_string_free(str, FALSE);
-				}
-				else
 					break;
+				}
 
 				j++;
 			}
```

File: src/tagmanager/tm_ctags.c (hunk 1 above)
Comment (techee) [reply src/tagmanager/tm_ctags.c:245]: Don't worry, I'm not going to use it as a base for some general-purpose string utility library ;-)

File: src/tagmanager/tm_ctags.c [hunk 4]
```diff
@@ -241,6 +241,29 @@ void tm_ctags_clear_ignore_symbols(void)
 }
 
 
+static void replace_str(gchar **where, const gchar *what, guint what_len,
+	const gchar *replacement, guint replacement_len)
+{
+	if (where && *where)
+	{
+		guint where_len = strlen(*where);
```
[the remark this answers is not in the document]
Comment (techee) [reply src/tagmanager/tm_ctags.c:249]: OK.

File: src/tagmanager/tm_ctags.c (hunk 3 above)
Comment (techee) [reply src/tagmanager/tm_ctags.c:255]: OK.

File: src/tagmanager/tm_ctags.c (hunk 2 above)
Comment (techee) [reply src/tagmanager/tm_ctags.c:393]: Right, I missed we are doing the same replacement once more below. We could maybe even use the GString version, I'm not sure if the performance is too critical here.

I checked what you propose and it looks good to me and unifies the code a little.

**Comment by techee** [comment]
@b4n Thanks for the review! I've just re-pushed the version with the changes you suggested.

File: src/tagmanager/tm_ctags.c [hunk 5]
```diff
@@ -241,6 +241,33 @@ void tm_ctags_clear_ignore_symbols(void)
...
+	const gchar *replacement, gsize replacement_len)
+{
+	if (where && *where)
+	{
+		gchar *pos = strstr(*where, what);
+
+		if (pos)
+		{
+			gsize where_len = strlen(*where);
+			gchar *str = g_malloc(where_len + (replacement_len - what_len) + 1);
```
Comment (b4n) [inline src/tagmanager/tm_ctags.c:254]: hum, the issue might be the parentheses here: as those are both unsigned, wouldn't it wrap around if `what_len < replacement_len`? if that's it, just dropping the parentheses should work (as `what_len` cannot be longer than `where_len`, and `replacement_len` cannot be negative)

Comment (b4n) [reply src/tagmanager/tm_ctags.c:254]: sorry to have created a mess in your PR…

**Comment by techee** [comment]
I had to make one more change, replace `guint` with `gsize` in `what_len` and `replacement_len`:
```C
static gboolean replace_str(gchar **where, const gchar *what, gsize what_len,
	const gchar *replacement, gsize replacement_len)
```
because otherwise
```C
gchar *str = g_malloc(where_len + (replacement_len - what_len) + 1);
```
may underflow.

Interestingly, only meson unit tests failed and not the autotools ones - not sure why.

File: src/tagmanager/tm_ctags.c (hunk 5 above)
Comment (techee) [reply src/tagmanager/tm_ctags.c:254]: Ah, I only noticed your message now. The gsize change I made should be alright too, right?

Comment (b4n) [reply src/tagmanager/tm_ctags.c:254]: I don't see why they would? Admittedly the tests passed, but I can't see why using a larger unsigned would solve the underflow… unless maybe it subsequently overflows in final computation, putting sign back in the right direction?
even if that's so, it's a bit too scary even for me :)

Comment (b4n) [reply src/tagmanager/tm_ctags.c:254]: …yeah that's what happens, so I'd rather just drop the parentheses if you're fine with it :)

Comment (techee) [reply src/tagmanager/tm_ctags.c:254]: Ah, OK,  gsize isn't signed...

Anyway, I just tried to put guints back in the function parameters, dropped the parentheses, and... the `nested_anon.c` test failed. Something somewhere still isn't right.

Comment (techee) [reply src/tagmanager/tm_ctags.c:254]: > Anyway, I just tried to put guints back in the function parameters, dropped the parentheses, and... the nested_anon.c test failed. Something somewhere still isn't right.

Nah, forget it, I modified the test locally, it works alright.

Comment (b4n) [reply src/tagmanager/tm_ctags.c]: ah, good!  I tested it locally and couldn't get it to fail without the parentheses, so I was starting to loose a lot of hairs over that :laughing:

Comment (techee) [reply src/tagmanager/tm_ctags.c]: Yeah, sorry for your hair loss :-)

File: src/tagmanager/tm_ctags.c [hunk 6]
```diff
@@ -241,6 +241,29 @@ void tm_ctags_clear_ignore_symbols(void)
...
+	const gchar *replacement, guint replacement_len)
+{
+	if (where && *where)
+	{
+		guint where_len = strlen(*where);
+		gchar *pos = strstr(*where, what);
+
+		if (pos)
+		{
+			gchar *str = g_malloc(where_len + replacement_len);
```
[the remark this answers is not in the document]
Comment (b4n) [reply src/tagmanager/tm_ctags.c]: …and hair pulling that arose from my change proved you right :)

**Pull request merged**